components/engine_test/src/lib.rs

```diff
@@ -0,0 +1,451 @@
+        path: &str,
+        db_opt: DBOptions,
+        cfs_opts: Vec<CFOptions>,
+    ) -> Result<RaftTestEngine> {
+        RaftTestEngine::new_engine_opt(path, db_opt, cfs_opts)
+    }
+}
+
+/// Types and constructors for the "kv" engine
+pub mod kv {
+    use crate::ctor::{CFOptions, DBOptions, EngineConstructorExt};
+    use engine_traits::Result;
+
+    #[cfg(feature = "test-engine-kv-panic")]
+    pub use engine_panic::{
+        PanicEngine as KvTestEngine, PanicSnapshot as KvTestSnapshot,
+        PanicWriteBatch as KvTestWriteBatch,
+    };
+
+    #[cfg(feature = "test-engine-kv-rocksdb")]
+    pub use engine_rocks::{
+        RocksEngine as KvTestEngine, RocksSnapshot as KvTestSnapshot,
+        RocksWriteBatch as KvTestWriteBatch,
+    };
+
+    pub fn new_engine(
+        path: &str,
+        db_opt: Option<DBOptions>,
+        cfs: &[&str],
+        opts: Option<Vec<CFOptions>>,
+    ) -> Result<KvTestEngine> {
+        KvTestEngine::new_engine(path, db_opt, cfs, opts)
+    }
+
+    pub fn new_engine_opt(
+        path: &str,
+        db_opt: DBOptions,
+        cfs_opts: Vec<CFOptions>,
+    ) -> Result<KvTestEngine> {
+        KvTestEngine::new_engine_opt(path, db_opt, cfs_opts)
+    }
+}
+
+/// Create a storage engine with a concrete type. This should ultimately be the
+/// only module within TiKv that needs to know about concrete engines. Other
+/// code only uses the `engine_traits` abstractions.
+///
+/// At the moment this has a lot of open-coding of engine-specific
+/// initialization, but in the future more constructor abstractions should be
+/// pushed down into engine_traits.
+///
+/// This module itself is intended to be extracted from this crate into its own
+/// crate, once the requirements for engine construction are better understood.
+pub mod ctor {
+    use engine_traits::Result;
+
+    /// Engine construction
+    ///
+    /// For simplicity, all engine constructors are expected to configure every
+    /// engine such that all of TiKV and its tests work correctly, for the
+    /// constructed column families.
+    ///
+    /// Specifically, this means that RocksDB constructors should set up
+    /// all properties collectors, always.
+    pub trait EngineConstructorExt: Sized {
+        /// Create a new engine with either:
+        ///
+        /// - The column families specified as `cfs`, with default options, or
+        /// - The column families specified as `opts`, with options.
+        ///
+        /// Note that if `opts` is not `None` then the `cfs` argument is completely ignored.
+        fn new_engine(
+            path: &str,
+            db_opt: Option<DBOptions>,
+            cfs: &[&str],
+            opts: Option<Vec<CFOptions>>,
+        ) -> Result<Self>;
+
+        /// Create a new engine with specified column families and options
+        fn new_engine_opt(path: &str, db_opt: DBOptions, cfs_opts: Vec<CFOptions>) -> Result<Self>;
+    }
+
+    #[derive(Clone)]
+    pub enum CryptoOpts {
```

Comment:
I think CryptoOptions would be a better name - its only a few chars and means it follows db options and cf options.

Reply:
Will do.